### PR TITLE
Add Datamodel and Django Model definitions

### LINKED ISSUE #4358: Django Schema Independence
Create static Django models so that the Specify 7 databases schema will no longer be reliant on the data_model.xml file from Specify 6.  In addition to providing independence for future development and schema updates, having static code, not generated at runtime, will create a better development environment.  This is an issue that would be best to solve before further development of the database schema

### PR DESCRIPTION
**Pull request by acwhite211**
Fixes #4358 and #2813

This PR adds datamodel and Django model code definitions, instead of relying on reading the 'specify_datamodel.xml' file from Specify 6.  So, the schema is now statically defined in our python Django code, instead of dynamically be created at runtime.  There is also code that can generated the static python code from the 'specify_datamodel.xml', so we can regenerate the code again in the future easily if we need to.  Unit tests have been created to ensure the equivalence  between the old and new datamodels.  There was another branch `django-schema` that had a lot of code from the development process, but I cleaned it up and simplified it into this branch.  There were some problems with adding the sql alchemy schema model classes, this isn't a pressing matter so this is something we can do in the future.

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions

Unfortunately there isn't anything specific to test.  I would say the best way to go about testing this from the UX side is to play around with forms and queries, and make sure no database table or field related errors occur.  Also, for the sp7_only tables, test them out in the query builder.

sp7_only tables:
- Spuserexternalid
- Spattachmentdataset
- UniquenessRule
- UniquenessRuleField
- Message
- Spmerging
- UserPolicy
- Role
- LibraryRole
- UserRole
- RolePolicy
- LibraryRolePolicy
- Spdataset

PR Testing Checklist:
- [x] QB works on tables like Agent, Collecting Object, and Locality
- [x] QB works on the new sp7_only tables
- [x] The Database Schema page works without errors (UserTools->DatabaseSchema) localhost/specify/data-model
- [x] Test out simple agent merging to make sure it still works properly

### CODE REVIEW

**Review by lexiclevenger — CHANGES_REQUESTED** [review]
### Testing instructions

Unfortunately there isn't anything specific to test.  I would say the best way to go about testing this from the UX side is to play around with forms and queries, and make sure no database table or field related errors occur.  Also, for the sp7_only tables, test them out in the query builder.

sp7_only tables:
- Spuserexternalid
- Spattachmentdataset
- UniquenessRule
- UniquenessRule_Field
- Message
- Spmerging
- UserPolicy
- Role
- LibraryRole
- UserRole
- RolePolicy
- LibraryRolePolicy
- Spdataset

PR Testing Checklist:
- [x] QB works on tables like Agent, Collecting Object, and Locality
- [ ] QB works on the new sp7_only tables
- [x] The Database Schema page works without errors (UserTools->DatabaseSchema) localhost/specify/data-model
---
Trying to open results from queries for the new tables results in a 404 error:

https://github.com/specify/specify7/assets/164079735/8dc49bc4-0c0f-456e-912f-7fb24d606e2f

File: specifyweb/specify/sp7_build_models.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    'java.util.Date': 'models.DateTimeField', # This mapping wasn't in the original code
+    'java.sql.Timestamp': 'models.DateTimeField',
+    'java.lang.Integer': 'models.IntegerField',
+    'java.lang.Long': 'models.BigIntegerField',
+    'java.lang.Byte': 'models.SmallIntegerField',
+    'java.lang.Short': 'models.SmallIntegerField',
+    'java.math.BigDecimal': 'models.DecimalField',
+    'java.lang.Double': 'models.FloatField', # This mapping wasn't in the original code
+    'java.lang.Boolean': 'models.BooleanField',
+    'text': 'models.TextField',
```
Comment (realVinayak) [inline specifyweb/specify/sp7_build_models.py:41]: Feel free to shot it down. there is a discrepancy in how datamodel defines spappresourcedata. It is actually a BLOB but treated as a TextField in django, which required ugly workaround in record merging. Could you look at schema (in mariadb) and figure out more cases like that, and possibly resolve them?

File: specifyweb/specify/models.py
Comment (maxpatiiuk) [inline specifyweb/specify/models.py:1]: why does both this file and datamodel.py exist?
are they creating 2 sources of truth now for where specify datamodel may be defined, and thus introducing possibility for inconsistency?
can this be avoided and have one place be derived from the other? or if not, then at very least, one of the files should be declared "off limits to edits by humans" and a script should be developed that would re-create one of these files from the other (or, recreate both from some third sourfe, like .XML or .json file)

File: specifyweb/specify/record_merging.py [hunk 2]
```diff
@@ -15,13 +15,31 @@
...
+from specifyweb.workbench import models as workbench_models
 from . import api, models as spmodels
 from .api import uri_for_model
 from .build_models import orderings
 from .load_datamodel import Table, FieldDoesNotExistError
 from celery.utils.log import get_task_logger # type: ignore
 logger = get_task_logger(__name__)
 
+
+APP_MODELS = [spmodels, acccounts_models, attachment_gw_models, businessrules_models, context_models,
```
Comment (maxpatiiuk) [inline specifyweb/specify/record_merging.py:34]: does APP_MODULES & get_app_model variable belong in this file, or should it be moved into some more central place?

File: specifyweb/specify/models.py
Comment (maxpatiiuk) [reply specifyweb/specify/models.py:1]: edit:
is specifyweb/specify/sp7_build_datamodel.py and sp7_build_models.py the files responsible for that?
if so, could we add a clear comment at the top of models.py and datamodel.py that they should not be edited by hand, with instructions (or link to) for how to regenerate these files and what each of these files means/what exactly is used for

for reference, see the datamodel types we have on the front-end:
https://github.com/specify/specify7/blob/43c4b7484498e32521e123591794ea06b4614dc1/specifyweb/frontend/js_src/lib/components/DataModel/types.ts#L1-L18
besides instructions, the file also includes last generated date and schema version

**Review by maxpatiiuk — APPROVED** [review]
lots of changes in this pr
this pr essentially means the "schema version" concept is now bound to "specify 7 version"
what are the advantages of statically defining the schema in python, rather than in some .json or .XML file? was that a pre-requisite for making back-end stateless?

to make specify's datamodel more flexible and customizable, if schema is determined by a .json/.xml file on the fly, we could have let (advanced) users customize the datamodel. now, for better or worse, this is no longer possible - worth the trade offs?

**Review by sharadsw — APPROVED** [review]

File: specifyweb/specify/sp7_build_models.py (hunk 1 above)
Comment (acwhite211) [reply specifyweb/specify/sp7_build_models.py:41]: Good idea, I'll look into that

**Comment by acwhite211** [comment]
@lexiclevenger, this PR can be re-tested now.  I made changes so that the form views of the sp7_only tables should work now.  Thanks

File: specifyweb/specify/record_merging.py (hunk 2 above)
Comment (acwhite211) [reply specifyweb/specify/record_merging.py:35]: ya, there might be a better place for this.

File: specifyweb/specify/sp7_build_models.py (hunk 1 above)
Comment (acwhite211) [reply specifyweb/specify/sp7_build_models.py:41]: I started work on the blob fields in this branch https://github.com/specify/specify7/tree/issue-4358-blobs, still a work in progress, most of the workbench upload unit tests failed.

**Comment by acwhite211** [comment]
I'm looking into this error that is happening during the GitHub Action test run.  Running the tests locally is working fine
```
hypothesis.errors.Flaky: Hypothesis
test_validate_upload_table_to_json(self=<specifyweb.workbench.upload.tests.testschema.OtherSchemaTests
testMethod=test_validate_upload_table_to_json>, name='10', wbcols={}) produces unreliable results: Falsified on the first
call but did not on a subsequent one
```

**Review by lexiclevenger — CHANGES_REQUESTED** [review]
PR Testing Checklist:
- [ ] QB works on tables like Agent, Collecting Object, and Locality
- [ ] QB works on the new sp7_only tables
- [ ] The Database Schema page works without errors (UserTools->DatabaseSchema) localhost/specify/data-model

---

Adding records to one of the forms opened by querying for the new tables causes an error.

1. Query for Uniqueness Rules (formatted)
2. Open one of the links
3. Click "Add," write in the modelname field, and save
4. Go back and query for the same table


https://github.com/specify/specify7/assets/164079735/b35ec2ed-2686-4cc9-94b1-fd08943c232f

This is KUHerps_2024_06_06. Also happened on KUFish_2024_06_06.


The error was not thrown right away for me, but eventually, a crash report appears every time I try to open the database:
[Specify 7 Crash Report - 2024-06-10T21_35_23.139Z.txt](https://github.com/user-attachments/files/15779319/Specify.7.Crash.Report.-.2024-06-10T21_35_23.139Z.txt)

**Comment by acwhite211** [comment]
Thanks @lexiclevenger, I was able to fix that issue with the UniquenessRule form.  I'll try to test out rest of the new forms to make sure the others work as well.  Should be good to re-test again.

**Comment by acwhite211** [comment]
@specify/testing, I added one extra step to the testing instructions "Test out simple agent merging to make sure it still works properly"

File: specifyweb/specify/sp7_build_models.py (hunk 1 above)
Comment (acwhite211) [reply specifyweb/specify/sp7_build_models.py:41]: ok, in this branch https://github.com/specify/specify7/tree/issue-4358-blobs, I changed the django fields from `TextField` to `BinaryField` where the mysql field is `mediumblob`.  I then added functions to override the save function for those fields to make sure to encode the data before saving it to the database.  I also added a `get_decoded_data` function as well.

@realVinayak do these changes suffice for avoiding issues in record merging?  Is there a unit test for this edge case?

**Review by Areyes42 — CHANGES_REQUESTED** [review]
### Testing instructions

Unfortunately there isn't anything specific to test. I would say the best way to go about testing this from the UX side is to play around with forms and queries, and make sure no database table or field related errors occur. Also, for the sp7\_only tables, test them out in the query builder.

sp7\_only tables:

-   Spuserexternalid
-   Spattachmentdataset
-   UniquenessRule
-   UniquenessRuleField
-   Message
-   Spmerging
-   UserPolicy
-   Role
-   LibraryRole
-   UserRole
-   RolePolicy
-   LibraryRolePolicy
-   Spdataset

PR Testing Checklist:

-   [x] QB works on tables like Agent, Collecting Object, and Locality
-   [ ] QB works on the new sp7\_only tables
-   [x] The Database Schema page works without errors (UserTools->DatabaseSchema) localhost/specify/data-model
-   [x] Test out simple agent merging to make sure it still works properly
***
I was able to recreate the issue with Spattachmentdataset as @emenslin mentioned.
Screenshot 2024-06-11 at 3 28 45 PM

Also, I get an 'Unable to find SpLocaleConatiner for Role' error when trying to access any of the sp7_only tables in Schema Config
Screenshot 2024-06-11 at 3 48 30 PM
[Specify 7 Crash Report - 2024-06-11T21_03_51.256Z.txt](https://github.com/user-attachments/files/15794841/Specify.7.Crash.Report.-.2024-06-11T21_03_51.256Z.txt)


Lastly, there is a small typo on the default label for description on LibraryRole
Screenshot 2024-06-11 at 4 19 19 PM

File: specifyweb/stored_queries/queryfieldspec.py [hunk 3]
```diff
@@ -179,6 +179,13 @@ def is_temporal(self):
         field = self.get_field()
         return field is not None and field.is_temporal()
 
+    def is_json(self):
+        field = self.get_field()
+        is_json_field = field is not None and field.type == 'json'
+        if field.sqlalchemy_type is not None: # hasattr(field, 'sqlalchemy_type') and
+            is_json_field = field.sqlalchemy_type.lower() == 'json'
+        return is_json_field
```
Comment (realVinayak) [inline specifyweb/stored_queries/queryfieldspec.py:187]: Hmm why do you need sqlalchemy_type?

File: specifyweb/stored_queries/format.py [hunk 4]
```diff
@@ -340,6 +340,9 @@ def _fieldformat(self, specify_field: Field,
             return cast(field,
                         types.Numeric(65))  # 65 is the mysql max precision
 
+        if isinstance(field.comparator.type, types.JSON):
```
Comment (realVinayak) [inline specifyweb/stored_queries/format.py:343]: won't the `specify_field.type` be "json" (so you won't have to add an additional sqlalchemy_type)?

File: specifyweb/stored_queries/queryfieldspec.py (hunk 3 above)
Comment (acwhite211) [reply specifyweb/stored_queries/queryfieldspec.py:187]: I initially added it to allow for a field to be 'json' in Django, and 'text' in sql.  I think this might be useful in the future, but if we end up not using it ever, then I'll remove it.

File: specifyweb/stored_queries/format.py (hunk 4 above)
Comment (acwhite211) [reply specifyweb/stored_queries/format.py:343]: hmm... `isinstance(field.comparator.type, types.JSON)` this is working.
I see what you mean `field.comparator.type` -> `JSON()`

I could do
```
from sqlalchemy.sql.sqltypes import JSON
isinstance(field.comparator.type, JSON)
```
but I don't really see that as any better of a solution

Comment (realVinayak) [reply specifyweb/stored_queries/format.py:343]: `specify_field`. that's the field defined in the datamodel (NOT django OR SQLAlchemy). It should be of type json. If not, make it.

**Comment by realVinayak** [comment]
^Could you add a unit test for the dataset? Too encourage it, it is not too complicated. just a couple of lines will do.
Here, `test_create_record_set` in `DataSetTests`, use the test_session_context. see usage `test_collection_object_count` for reference.

```
with test_session_context() as session:
     query = session.query(models.Spdataset.data).limit(1) # this is SQL model (not django), do the import approriately
     assert list(query) = ["[["1"], ["2"], ["3"]]"]
```

that's it^

**Comment by grantfitzsimmons** [comment]
Some possible table descriptions (feel free to make edits for accuracy/clarification):

---

*SpUserExternalId*
Stores provider identifiers and tokens for users who sign in using Single Sign On (SSO).

*SpAttachmentDataSet*
Holds attachment data sets.

*UniquenessRule*
Stores table names in the data model that have uniqueness rules configured for each discipline.

*UniquenessRuleField*
Stores field names in the data model that have uniqueness rules configured for each discipline, linked to UniquenessRule records.

*NotificationsMessage*
Stores user notifications.

*SpMerging*
Tracks record and task IDs of records being merged.

*SpUserPolicy*
Records permissions for a user within a collection.

*SpUserRole*
Records roles associated with Specify users.

*SpRole*
Stores names, descriptions, and collection information for user-created roles.

*SpRolePolicy*
Stores resource and action permissions for user-created roles within a collection.

*SpLibraryRole*
Stores names and descriptions of default roles that can be added to any collection.

*SpLibraryRolePolicy*
Stores resource and action permissions for library roles within a collection.

*SpDataSet*
Stores Specify Data Sets created during bulk import using the WorkBench, typically through spreadsheet uploads.

---

Something to consider– we probably shouldn't let anyone aside from institution admins edit the `spuser*` tables to prevent full access users from giving themselves administrative permissions, even if they have permission to create, edit, delete in all tables.

**Review by lexiclevenger — APPROVED** [review]
PR Testing Checklist:
- [x] QB works on tables like Agent, Collecting Object, and Locality
- [x] QB works on the new sp7_only tables
- [x] The Database Schema page works without errors (UserTools->DatabaseSchema) localhost/specify/data-model
- [x] Test out simple agent merging to make sure it still works properly
---
- [x]  Spattachmentdataset QB fix data field
- [x]  Description typo in LibraryRole
---
Looks great!

**Review by emenslin — APPROVED** [review]
### Testing instructions

Unfortunately there isn't anything specific to test.  I would say the best way to go about testing this from the UX side is to play around with forms and queries, and make sure no database table or field related errors occur.  Also, for the sp7_only tables, test them out in the query builder.

sp7_only tables:
- Spuserexternalid
- Spattachmentdataset
- UniquenessRule
- UniquenessRuleField
- Message
- Spmerging
- UserPolicy
- Role
- LibraryRole
- UserRole
- RolePolicy
- LibraryRolePolicy
- Spdataset

PR Testing Checklist:
- [x] QB works on tables like Agent, Collecting Object, and Locality
- [x] QB works on the new sp7_only tables
- [x] The Database Schema page works without errors (UserTools->DatabaseSchema) localhost/specify/data-model
- [x] Test out simple agent merging to make sure it still works properly

Everything within this pr seems to be working great!

**Review by Areyes42 — APPROVED** [review]
### Testing instructions

PR Testing Checklist:

-   [x] QB works on tables like Agent, Collecting Object, and Locality
-   [x] QB works on the new sp7\_only tables
-   [x] The Database Schema page works without errors (UserTools->DatabaseSchema) localhost/specify/data-model
-   [x] Test out simple agent merging to make sure it still works properly
-  [x]  Spattachmentdataset QB fix data field
-  [x]  Description typo in LibraryRole
***
Retested, everything looks good now! 🤸
